<a href="https://colab.research.google.com/github/angelbastogz/masters-ai/blob/main/Programaci%C3%B3n%20para%20Inteligencia%20Artificial/01_Python_Moderno/Notebooks/u1_n1_curso_acelerado_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Curso acelerado de Python para IA

**Unidad 1 · Semana 1 · Notebook 1 de 2**

Bienvenido a este taller para aprender a representar datos, transformarlos y construir objetos con estado.
No se presupone experiencia en Python ni en programación.

Trabajaremos en [Google Colab](https://colab.research.google.com/) al inicio de las lecciones. El curso utiliza Python 3.12 o
posterior.


## Objetivos y forma de trabajo

Al terminar podrás elegir entre listas, tuplas, diccionarios y conjuntos; escribir
funciones pequeñas; interpretar errores; y construir una clase cuyos objetos
mantengan estados independientes.

La ruta tiene tres tramos:

1. **Primeros pasos:** valores, cadenas, colecciones, decisiones y ciclos.
2. **Taller de datos:** funciones, conteo y agrupación de registros.
3. **POO:** instancias, atributos, métodos y un registro de predicciones.

En cada ejercicio: **predice → ejecuta → explica → modifica**. Equivocarse y leer el
error forma parte del trabajo. Cada ejercicio debe resolverse en su celda y conservar
su explicación y sus comprobaciones junto al código. Las actividades de
profundización permiten ampliar el trabajo de clase.


## 0. Antes de comenzar

En Colab, una **celda de código** conserva sus variables después de ejecutarse.
Puedes ejecutar una celda con `Shift + Enter`. Si el estado de la sesión actual se deteriora,
usa **Entorno de ejecución → Reiniciar sesión** y ejecuta de nuevo desde arriba.

In [23]:
import platform
import sys

print(f"Python: {sys.version.split()[0]}")
print(f"Plataforma: {platform.platform()}")

Python: 3.13.15
Plataforma: Linux-6.6.122+-x86_64-with-glibc2.39


## 1. Valores, variables y tipos

Python asigna nombres a objetos. El nombre no queda ligado para siempre a un
tipo, por eso conviene elegir nombres que expresen intención y evitar cambios de
significado a mitad del programa.

In [24]:
modelo = "clasificador de reseñas"
precision = 0.91
ejemplos = 1_250
esta_entrenado = True
sin_descripcion = None

for nombre, valor in {
    "modelo": modelo,
    "precision": precision,
    "ejemplos": ejemplos,
    "esta_entrenado": esta_entrenado,
    "sin_descripcion": sin_descripcion,
}.items():
    print(f"{nombre:16} -> {valor!r:28} ({type(valor).__name__})")

modelo           -> 'clasificador de reseñas'    (str)
precision        -> 0.91                         (float)
ejemplos         -> 1250                         (int)
esta_entrenado   -> True                         (bool)
sin_descripcion  -> None                         (NoneType)


### Operadores esenciales

- Aritméticos: `+`, `-`, `*`, `/`, `//`, `%`, `**`.
- Comparación: `==`, `!=`, `<`, `<=`, `>`, `>=`.
- Lógicos: `and`, `or`, `not`.
- Identidad: `is` y `is not`, especialmente con `None`.

`=` asigna; `==` compara valores. Para preguntar si falta un valor escribimos
`valor is None`, no `valor == None`.

In [25]:
aciertos = 91
total = 100

proporcion = aciertos / total
porcentaje = proporcion * 100
cumple_objetivo = proporcion >= 0.90

print(proporcion, porcentaje, cumple_objetivo)

0.91 91.0 True


### Cadenas y f-strings

Las cadenas son secuencias inmutables. Las f-strings permiten insertar y
formatear valores sin construir texto mediante concatenaciones difíciles de leer.

In [26]:
texto = "  La IA necesita datos claros.  "
limpio = texto.strip()
palabras = limpio.lower().split()

print(limpio)
print(palabras)
print(f"El texto contiene {len(palabras)} palabras y {len(limpio)} caracteres.")
print(f"Precisión reportada: {precision:.1%}")

La IA necesita datos claros.
['la', 'ia', 'necesita', 'datos', 'claros.']
El texto contiene 5 palabras y 28 caracteres.
Precisión reportada: 91.0%


### Cómo leer un programa

Una instrucción indica una acción; una expresión produce un valor. Python ejecuta
las instrucciones de arriba hacia abajo. `print(...)` muestra un valor, `type(...)`
indica su tipo y `#` inicia un comentario. Los paréntesis permiten llamar funciones.

En el primer ejemplo usamos un `for` para mostrar varios valores: más adelante lo
construiremos paso a paso. Por ahora modifica un valor y observa qué cambia.

### Ejercicio 1 · Una frase con datos

Crea las variables `nombre_modelo`, `exactitud` y `cantidad_pruebas`. Muestra `modelo-x obtuvo 87.5% en 80 pruebas`. Después cambia la exactitud a cero: ¿debería interpretarse como un dato ausente?

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [27]:
# Define tus variables y construye una f-string.
nombre_modelo = "modelo-x"
exactitud = 0.875
cantidad_pruebas = 80
print(f"{nombre_modelo} obtuvo {exactitud:.1%} en {cantidad_pruebas} pruebas.")

exactitud = 0.0
print(f"{nombre_modelo} obtuvo {exactitud:.1%} en {cantidad_pruebas} pruebas.")

modelo-x obtuvo 87.5% en 80 pruebas.
modelo-x obtuvo 0.0% en 80 pruebas.


## 2. Colecciones

Elegir una colección es una decisión de diseño:

| Tipo | Orden | Mutable | Uso habitual |
|---|---|---|---|
| `list` | Sí | Sí | Secuencia que cambia |
| `tuple` | Sí | No | Registro corto o valor compuesto |
| `dict` | Sí* | Sí | Asociación clave → valor |
| `set` | No significativo | Sí | Valores únicos y pertenencia rápida |

Nota: los diccionarios conservan el orden de inserción, pero se consultan por clave.

In [28]:
etiquetas = ["positivo", "neutral", "negativo", "positivo"]
coordenada = (20.97, -89.62)
prediccion = {"texto": "Excelente servicio", "etiqueta": "positivo", "score": 0.97}
etiquetas_unicas = set(etiquetas)

print(etiquetas[0], etiquetas[-1], etiquetas[1:3])
print(coordenada)
print(prediccion["etiqueta"], prediccion.get("idioma", "desconocido"))
print(etiquetas_unicas)

positivo positivo ['neutral', 'negativo']
(20.97, -89.62)
positivo desconocido
{'negativo', 'neutral', 'positivo'}


### Listas: posiciones y operaciones

El primer índice es `0`; `-1` indica el último elemento. Un corte `[inicio:fin]`
incluye el inicio y excluye el fin. `append` agrega un elemento y `extend` agrega
los elementos de otra colección. Ambos modifican la lista y devuelven `None`.

In [29]:
muestras = ["a", "b", "c", "d"]
print(muestras[:2], muestras[2:], muestras[::2])
muestras.append("e")
muestras.extend(["f", "g"])
print(muestras)

['a', 'b'] ['c', 'd'] ['a', 'c']
['a', 'b', 'c', 'd', 'e', 'f', 'g']


### Ejercicio 2 · Una lista dentro de otra

Predice la diferencia entre `append` y `extend` antes de ejecutar. Después construye una lista plana con los cuatro textos y extrae los dos últimos usando un corte.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [30]:
lote_a = ["hola", "gracias"]
lote_b = ["adiós", "hasta luego"]
con_append = lote_a.copy()
con_extend = lote_a.copy()
# Aplica append y extend por separado e inspecciona los resultados.
con_append.append(lote_b)
con_extend.extend(lote_b)
print(con_append)
print(con_extend)

['hola', 'gracias', ['adiós', 'hasta luego']]
['hola', 'gracias', 'adiós', 'hasta luego']


### Mutabilidad, alias y copia

Dos nombres pueden apuntar a la misma lista. Para modificar el contenedor por separado, debemos copiarlo.
Si contiene otros objetos mutables, una copia superficial aún puede compartirlos. Este detalle explica muchos errores silenciosos
en pipelines de datos.

In [31]:
original = ["dato-a", "dato-b"]
alias = original
copia = original.copy()

alias.append("dato-c")
copia.append("dato-independiente")

print("original:", original)
print("alias:   ", alias)
print("copia:   ", copia)
print("¿Mismo objeto?", original is alias, original is copia)

original: ['dato-a', 'dato-b', 'dato-c']
alias:    ['dato-a', 'dato-b', 'dato-c']
copia:    ['dato-a', 'dato-b', 'dato-independiente']
¿Mismo objeto? True False


### Copiar una estructura anidada

`.copy()` hace una **copia superficial**: crea el contenedor externo, pero conserva
las referencias a sus elementos. Si dentro hay una lista, esa lista sigue compartida.

In [32]:
registro_original = {"texto": "hola", "etiquetas": ["saludo"]}
registro_copia = registro_original.copy()
registro_copia["etiquetas"].append("corto")
print(registro_original)
print(registro_original is registro_copia)
print(registro_original["etiquetas"] is registro_copia["etiquetas"])

{'texto': 'hola', 'etiquetas': ['saludo', 'corto']}
False
True


### Ejercicio 3 · Copiar sin compartir la lista

Crea una copia que permita añadir `corto` sin cambiar las etiquetas del original. Para esta estructura basta copiar también la lista interna. **Profundización:** investiga qué cambiaría si hubiera más niveles.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [33]:
from IPython.core.display import deepcopy
fuente = {"texto": "hola", "etiquetas": ["saludo"]}
# Construye una copia independiente para esta estructura.
fuente_copia = fuente.copy()
fuente_copia["etiquetas"] = fuente["etiquetas"].copy()
fuente["etiquetas"].append("corto")
print(fuente)
print(fuente_copia)

{'texto': 'hola', 'etiquetas': ['saludo', 'corto']}
{'texto': 'hola', 'etiquetas': ['saludo']}


### Tuplas: agrupar y desempaquetar

Una tupla permite reunir valores en posiciones conocidas. No podemos reemplazar
sus elementos. Al desempaquetar, la cantidad de nombres debe coincidir con la de valores.

In [34]:
evaluacion = ("modelo-a", 0.87)
nombre, score = evaluacion
print(nombre, score)
# Una tupla puede contener una lista mutable: inmutable no significa congelar todo.

modelo-a 0.87


### Diccionarios: representar registros

Un diccionario asocia claves a valores. `in` comprueba las claves; `.items()`
permite recorrer pares clave–valor. `get` usa su valor por defecto cuando la clave
no existe, pero conserva `None` si ese es el valor guardado.

In [35]:
ficha = {"texto": "Excelente", "score": None}
ficha["etiqueta"] = "positivo"
print("score" in ficha)
print(ficha.get("score", 0), ficha.get("idioma", "es"))
for clave, valor in ficha.items():
    print(clave, valor)

True
None es
texto Excelente
score None
etiqueta positivo


### Conjuntos: pertenencia, coincidencias y diferencias

Un conjunto conserva valores únicos. Usa `set()` para el conjunto vacío: `{}` crea
un diccionario. La intersección `&` encuentra coincidencias, la unión `|` reúne
valores y la diferencia `-` encuentra los que están sólo en el primer conjunto.
No uses la posición ni el orden impreso de un conjunto como parte del resultado.

In [36]:
permitidas = {"positivo", "neutral", "negativo"}
observadas = {"positivo", "positivo", "desconocido"}
print("Reconocidas:", observadas & permitidas)
print("No reconocidas:", observadas - permitidas)

Reconocidas: {'positivo'}
No reconocidas: {'desconocido'}


### Ejercicio 4 · Elegir una estructura

Representa dos predicciones con texto, etiqueta y score. Conserva el orden de llegada y encuentra las etiquetas distintas. Explica por qué conviene una lista de diccionarios para los registros y un conjunto para las etiquetas.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [37]:
# Crea dos registros, guárdalos en una lista y reúne sus etiquetas en un set.
predicciones = [
    {"texto": "Excelente", "etiqueta": "neutral", "score": 0.98},
    {"texto": "Regular", "etiqueta": "neutral", "score": 0.66}
]
etiquetas = set()
for registro in predicciones:
    etiquetas.add(registro["etiqueta"])
print(etiquetas)

{'neutral'}


## 3. Decisiones y truthiness

En una condición, Python interpreta como falsos `False`, `None`, cero y las
colecciones vacías. Esto permite escribir condiciones directas, siempre que el
significado sea claro.

In [38]:
respuesta_modelo = ""

if not respuesta_modelo:
    estado = "sin respuesta"
elif len(respuesta_modelo) < 20:
    estado = "respuesta breve"
else:
    estado = "respuesta disponible"

print(estado)

sin respuesta


### Bloques, ciclos y decisiones

Los dos puntos abren un bloque y la sangría indica qué instrucciones pertenecen a
él. Usa cuatro espacios por nivel. `for` repite el bloque para cada elemento;
`range(3)` produce los valores 0, 1 y 2. `continue` salta al siguiente elemento y
`break` termina el ciclo. `while` repite mientras una condición sea verdadera.

In [39]:
for valor in [0.0, None, 0.85]:
    if valor is None:
        continue
    print("Score presente:", valor)

intentos = 0
while intentos < 3:
    intentos += 1
    print("Intento", intentos)

Score presente: 0.0
Score presente: 0.85
Intento 1
Intento 2
Intento 3


### Ejercicio 5 · Cero no es ausencia

Recorre `[None, 0.0, 0.4, 0.95, 0.99]`. Ignora `None` y detente en el primer score mayor o igual a 0.9. Deben mostrarse 0.0, 0.4 y 0.95. ¿Qué se perdería usando `if not score`?

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [40]:
valores_busqueda = [None, 0.0, 0.4, 0.95, 0.99]
# Usa for, continue y break.
for score in valores_busqueda:
    if score is None:
        continue
    if score >= 0.9:
        break
    print(score)

print("===")
for score in valores_busqueda:
    if not score:
        continue
    print(score)
    if score >= 0.9:
        break


0.0
0.4
===
0.4
0.95


## 4. Ciclos y comprehensions

Un `for` recorre los elementos de un iterable. En una lista puedes recorrer los
valores directamente; no hace falta construir sus índices. Usa `enumerate` cuando
también necesites la posición y `zip` para avanzar por varias secuencias a la vez.
Por defecto, `zip` se detiene al agotarse la secuencia más corta. Si necesitas
comprobar que tengan la misma longitud, usa `zip(..., strict=True)`.

In [41]:
textos = ["muy bueno", "regular", "excelente", "perfecto"]
scores = [0.82, 0.51, 0.96, 1]

for posicion, (texto, score) in enumerate(zip(textos, scores, strict=True), start=1):
    print(f"{posicion}. {texto:<12} → {score:.0%}")

1. muy bueno    → 82%
2. regular      → 51%
3. excelente    → 96%
4. perfecto     → 100%


Las comprehensions sirven para transformaciones pequeñas y legibles. Si la
lógica requiere varias decisiones o efectos secundarios, un ciclo explícito
suele ser mejor.

In [42]:
scores_altos = [score for score in scores if score >= 0.80]
longitudes = {texto: len(texto) for texto in textos}
iniciales = {texto[0] for texto in textos}

print(scores_altos)
print(longitudes)
print(iniciales)

[0.82, 0.96, 1]
{'muy bueno': 9, 'regular': 7, 'excelente': 9, 'perfecto': 8}
{'e', 'm', 'r', 'p'}


### Ejercicio 6 · Transformar y filtrar

Convierte a Fahrenheit las temperaturas mayores o iguales a 25. Usa la fórmula `fahrenheit = celsius * 9 / 5 + 32`. Primero usa un ciclo y después una comprehension. El resultado debe ser `[77.0, 84.2, 87.8]`.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [43]:
temperaturas = [18, 21, 25, 29, 31]
# Construye ambas versiones.
for temperatura in temperaturas:
    if temperatura >= 25:
        print(temperatura * 9 / 5 + 32)

fahrenheit = [temperatura * 9 / 5 + 32 for temperatura in temperaturas if temperatura >= 25]
print(fahrenheit)

77.0
84.2
87.8
[77.0, 84.2, 87.8]


## 5. Funciones: nombrar una transformación

Una función bien enfocada reduce repetición y permite probar una idea de forma
aislada. Sus parámetros son entradas; `return` produce una salida. Una función
sin `return` explícito devuelve `None`.

In [44]:
def normalizar_texto(texto, *, minusculas=True):
    '''Elimina espacios externos y opcionalmente convierte a minúsculas.'''
    resultado = texto.strip()
    if minusculas:
        resultado = resultado.lower()
    return resultado


print(normalizar_texto("  Hola, IA  "))
print(normalizar_texto("  Hola, IA  ", minusculas=False))

hola, ia
Hola, IA


El `*` vuelve explícitos los argumentos posteriores: debemos escribir
`minusculas=False`. En la semana 2 añadiremos anotaciones de tipo y Pydantic;
hoy nos concentramos en el comportamiento.

### Verificar una función con `assert`

`assert condición` comprueba una expectativa. Si es falsa, produce un
`AssertionError`. Lo usaremos para comprobar ejercicios; las entradas externas
se validarán con condiciones y excepciones explícitas.

In [45]:
assert normalizar_texto("  HOLA  ") == "hola"
assert normalizar_texto("  HOLA  ", minusculas=False) == "HOLA"
assert normalizar_texto("") == ""

### Ejercicio 7 · Promedio sin inventar datos

Escribe `promedio_o_none(valores)`: devuelve el promedio o `None` si la lista está vacía. Comprueba `[0.0]`, `[]` y `[0.5, 1.0]`. ¿Por qué cero y ausencia necesitan resultados distintos?

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [60]:
# Define promedio_o_none y escribe tres comprobaciones.
def promedio_o_none(valores):
    if not valores:
        return None
    return sum(valores) / len(valores)

assert promedio_o_none([0.0]) == 0.0
assert promedio_o_none([]) is None

### Ordenar usando una función

`sorted` crea una lista nueva; `.sort()` modifica la lista existente. El argumento
`key` recibe una función que calcula el criterio de orden. Pasamos la función
sin llamarla: Python la aplicará a cada elemento.

In [47]:
evaluaciones = [
    {"modelo": "a", "score": 0.7},
    {"modelo": "b", "score": 0.95},
    {"modelo": "c", "score": 0.8},
]

def obtener_score(evaluacion):
    return evaluacion["score"]

ranking = sorted(evaluaciones, key=obtener_score, reverse=True)
print(ranking)
# lambda permite expresar una función pequeña en el mismo lugar.
assert ranking == sorted(evaluaciones, key=lambda e: e["score"], reverse=True)

[{'modelo': 'b', 'score': 0.95}, {'modelo': 'c', 'score': 0.8}, {'modelo': 'a', 'score': 0.7}]


## 6. Módulos e imports

Un módulo es un archivo Python reutilizable. La biblioteca estándar ya contiene
muchas herramientas; conviene conocerla antes de añadir dependencias externas.
Importamos el módulo o nombres específicos y evitamos `from modulo import *`.

In [48]:
from collections import Counter
from statistics import mean

conteos = Counter(etiquetas)

print("Frecuencias:", conteos)
print("Más común:", conteos.most_common(1))
print("Score medio:", mean(scores))

Frecuencias: Counter({'neutral': 1})
Más común: [('neutral', 1)]
Score medio: 0.8225


### De contar a agrupar

Un contador responde «¿cuántos?». Una agrupación conserva los elementos asociados
a cada clave. `defaultdict(list)` crea una lista al acceder a una clave ausente.

In [62]:
from collections import defaultdict

conteo_manual = {}
for etiqueta in etiquetas:
    conteo_manual[etiqueta] = conteo_manual.get(etiqueta, 0) + 1
assert conteo_manual == dict(Counter(etiquetas))

por_etiqueta = defaultdict(list)
for registro in [
    {"etiqueta": "positivo", "score": 0.9},
    {"etiqueta": "neutral", "score": 0.6},
    {"etiqueta": "positivo", "score": 0.8},
]:
    por_etiqueta[registro["etiqueta"]].append(registro["score"])

print(dict(por_etiqueta))

{'positivo': [0.9, 0.8], 'neutral': [0.6]}


### Ejercicio 8 · Resumir grupos

Usando `por_etiqueta`, construye un diccionario con el promedio de cada grupo. Comprueba que positivo tiene dos scores y neutral uno. **Profundización:** ordena las etiquetas por promedio descendente.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [69]:
# Recorre por_etiqueta.items() y calcula cada promedio.
promedios = {}
items = por_etiqueta.items()
for etiqueta, scores in items:
    promedios[etiqueta] = sum(scores) / len(scores)
print(promedios)

{'positivo': 0.8500000000000001, 'neutral': 0.6}


## 7. Excepciones: leer antes de corregir

Cuando Python no puede continuar, muestra un **traceback**. Léelo desde la última
línea:

1. tipo de excepción (`ValueError`, `KeyError`, `TypeError`, etc.);
2. mensaje concreto;
3. línea de nuestro código donde ocurrió;
4. cadena de llamadas que llevó hasta allí.

Captura sólo excepciones que puedas manejar. Un `except:` sin tipo de excepción captura incluso interrupciones de la sesión.
Si además su cuerpo solo contiene `pass`, silencia el problema y dificulta el
diagnóstico. Un bloque literalmente vacío no es sintaxis válida.

### Laboratorio de error

Quita el comentario de **una línea a la vez**, ejecútala y explica el traceback
con tus propias palabras. Después vuelve a comentarla para que **Ejecutar todo**
pueda completar la notebook.

In [74]:
# int("tres")
# ["a", "b"][5]
# {"modelo": "demo"}["score"]
# "score=" + 0.91

## 8. POO: objetos que conservan estado

Una **clase** define cómo construir objetos y qué operaciones ofrecen. Cada
**instancia** es un objeto concreto. Sus **atributos** guardan datos y sus
**métodos** son funciones asociadas al objeto.

`__init__` inicializa una instancia; `self` representa esa instancia dentro de los
métodos. Al llamar `contador.agregar()`, Python pasa `contador` como `self`.

In [77]:
class Contador:
    def __init__(self, nombre):
        self.nombre = nombre
        self.total = 0

    def agregar(self):
        self.total += 1

    def reiniciar(self):
        self.total = 0

contador_a = Contador("lote A")
contador_b = Contador("lote B")
contador_a.agregar()
contador_a.agregar()
print(contador_a.nombre, contador_a.total)
print(contador_b.nombre, contador_b.total)
assert contador_b.total == 0

lote A 2
lote B 0


### Ejercicio 9 · Instancias y alias

Antes de ejecutar, predice qué ocurrirá al hacer `otro_nombre = contador_a` y llamar `otro_nombre.reiniciar()`. ¿Crea una instancia nueva? Luego crea un tercer contador y comprueba que agregarle un evento no cambia los otros.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [80]:
# Prueba el alias y una instancia nueva; explica sus diferencias.
otro_nombre = contador_a
otro_nombre.reiniciar()
print(otro_nombre.nombre, otro_nombre.total)
print(contador_a.nombre, contador_a.total)

contador_c = Contador("lote C")
contador_c.agregar()
print(contador_c.nombre, contador_c.total)

lote A 0
lote A 0
lote C 1


### Laboratorio POO: una lista compartida por accidente

Un atributo definido en el cuerpo de la clase se comparte entre las instancias.
Los datos propios de cada instancia deben inicializarse en `__init__`.

In [81]:
class RegistroCompartido:
    registros = []  # Defecto intencional: la lista pertenece a la clase.

    def agregar(self, texto):
        self.registros.append(texto)

compartido_a = RegistroCompartido()
compartido_b = RegistroCompartido()
compartido_a.agregar("dato de A")
print(compartido_b.registros)
assert compartido_a.registros is compartido_b.registros

['dato de A']


### Ejercicio 10 · Estado independiente

Crea `RegistroIndependiente` con una lista por instancia y un método `agregar`. Comprueba que agregar un dato a A deja vacía la lista de B.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.

In [82]:
# Mueve la inicialización de la lista a __init__.
class RegistroIndependiente:
    def __init__(self):
        self.registros = []

    def agregar(self, texto):
        self.registros.append(texto)

independiente_a = RegistroIndependiente()
independiente_b = RegistroIndependiente()
independiente_a.agregar("dato de A")
print(independiente_b.registros)
assert independiente_a.registros is not independiente_b.registros

[]


## 9. Caso integrador: un registro de predicciones

La clase siguiente conserva predicciones ya normalizadas y combina estructuras de
datos con comportamiento y estado. Cada registro recibe su posición al agregarse.
`resumen` devuelve cantidad, conteos y promedio; el promedio es `None` cuando no
hay registros.


In [83]:
predicciones = [
    {"texto": "Excelente", "etiqueta": "positivo", "score": 0.98},
    {"texto": "Regular", "etiqueta": "neutral", "score": 0.66},
    {"texto": "Llegó roto", "etiqueta": "negativo", "score": 0.93},
    {"texto": "Cero válido", "etiqueta": "neutral", "score": 0.0},
]


In [92]:
class RegistroPredicciones:
    def __init__(self, nombre):
        self.nombre = nombre
        self._registros = []

    def agregar(self, registro, posicion):
        guardado = registro.copy()
        guardado["posicion"] = posicion
        self._registros.append(guardado)

    def average_by_group(self):
        por_etiqueta = defaultdict(list)
        for registro in self._registros:
            por_etiqueta[registro["etiqueta"]].append(registro["score"])
        return {etiqueta: mean(scores) for etiqueta, scores in por_etiqueta.items()}

    def resumen(self):
        scores = [registro["score"] for registro in self._registros]
        return {
            "cantidad": len(self._registros),
            "conteos": dict(Counter(r["etiqueta"] for r in self._registros)),
            "promedio": sum(scores) / len(scores) if scores else None,
        }

    def reiniciar(self):
      self._registros = []


lote = RegistroPredicciones("Demostración")
for posicion, registro in enumerate(predicciones):
    lote.agregar(registro, posicion)
print(lote.resumen())


{'cantidad': 4, 'conteos': {'positivo': 1, 'neutral': 2, 'negativo': 1}, 'promedio': 0.6425000000000001}


### Leer el resultado antes de extenderlo

Deben quedar **4 registros**, conteos de positivo: 1, neutral: 2 y negativo: 1,
y promedio **0.6425**. Este promedio describe los scores; no mide exactitud del
modelo porque no tenemos etiquetas de referencia.


### Ejercicio 11 · Comprobar el contrato

Comprueba el resultado del lote con `assert`. Crea además un lote vacío: su promedio
debe ser `None`. Verifica que el registro con score cero se conserve.

Resuelve el ejercicio en la celda siguiente y conserva las comprobaciones.


In [89]:
# Comprueba cantidades, conteos, promedio, posiciones y ausencia de datos.
assert lote.resumen() == {
    "cantidad": 4,
    "conteos": {"positivo": 1, "neutral": 2, "negativo": 1},
    "promedio": 0.6425000000000001,
}

vacio = RegistroPredicciones("Vacio")
assert vacio.resumen() == {
    "cantidad": 0,
    "conteos": {},
    "promedio": None,
}

### Profundización

- Añade un método que agrupe scores por etiqueta y devuelva el promedio por grupo.
- Añade `reiniciar()` y comprueba que vacía válidas y rechazos de una sola instancia.
- Compara una función que procesa un lote completo con una instancia que recibe
  registros en distintas llamadas. ¿En qué situación te ayuda conservar estado?

La herencia y las jerarquías de clases quedan para una ampliación posterior.
Primero necesitamos dominar instancias, métodos y estado independiente.

## Cierre

Explica con un ejemplo cuándo elegirías cada estructura de datos, cómo distinguirías
cero de ausencia y qué ocurre al copiar un contenedor con elementos mutables.
Después identifica los atributos y métodos de `RegistroPredicciones` y explica
por qué dos lotes deben conservar listas distintas.


In [91]:
lote.average_by_group()

{'positivo': 0.98, 'neutral': 0.33, 'negativo': 0.93}

In [95]:
lote.reiniciar()
lote.resumen()

{'cantidad': 0, 'conteos': {}, 'promedio': None}